src/rewards/mod.rs

```diff
@@ -171,73 +173,87 @@ impl<'a> RewardsMonitor<'a> {
                 .into_iter()
                 .map(|(p, a)| ((p, current_epoch), a)),
         );
 
-        if !to_query.is_empty() {
-            // Create empty hashmap
-            let mut pka: HashMap<(_, _), _> = to_query
-                .iter()
-                .map(|pubkey| ((*pubkey, current_epoch), None))
-                .collect::<HashMap<_, _>>();
-
-            // Chunk into 100
-            for chunk in to_query.chunks(100) {
-                let account_infos = self.client.get_multiple_accounts(&chunk)?;
-
-                // Write to hashmap
-                // FIXME: Insert only APY data, calculate them here.
-                for (pubkey, account_info) in chunk.iter().zip(account_infos) {
-                    pka.insert((*pubkey, current_epoch), account_info);
-                }
-
-                let insert = pka
-                    .clone()
-                    .into_iter()
-                    .map(|((pk, _), a)| (pk, a))
-                    .collect::<HashMap<_, _>>();
+        if to_query.is_empty() {
+            return Ok(HashSet::new());
+        }
 
-                // Write to cache in chunks of 100 at a time.
-                // FIXME: This writes all account state into database, which is inefficient. Refactor so that only APY data is written and `accounts` only stores APY data and drop account information here.
-                self.cache.add_epoch_data(current_epoch, insert)?;
+        let mut pka = HashMap::new();
+
+        // Seen voters are added here so that an APY calculation occurs is done only once
+        // for a given voter.
+        let mut seen_voters = BTreeSet::new();
+        // Chunk into 100
+        for chunk in to_query.chunks(100) {
+            let pubkeys: Vec<_> = chunk.iter().map(|r| r.pubkey).collect();
+            let account_infos = self.client.get_multiple_accounts(pubkeys.as_slice())?;
+
+            // Write to hashmap
+            for (reward, account_info) in chunk.iter().zip(account_infos) {
+                if let Some(account_info) = account_info {
+                    let StakingApy { voter, percent } = calculate_staking_apy(
+                        &account_info,
+                        &mut seen_voters,
+                        3.0, /* FIXME: calculate */
+                        reward.lamports as u64,
+                        reward.post_balance,
+                    )?;
+                    pka.insert((voter, current_epoch), percent);
+                }
             }
 
-            // Extend accounts
-            accounts.extend(pka);
+            let insert = pka
+                .clone()
+                .into_iter()
+                .map(|((pk, _), a)| (pk, a))
+                .collect();
+
+            // Write to cache in chunks of 100 at a time.
+            self.cache.add_epoch_data(current_epoch, insert)?;
         }
 
+        // Extend accounts
+        apys.extend(pka);
+
         todo!("missing staking apy calculation using multiple epochs")
```

Comment:
This should be removed when the function is complete?

Reply:
Actually, I think this todo should stay until we can calculate APY across multiple years.